### PR TITLE
Restructure code to split mask calculation from rendering process

### PR DESCRIPTION
**Pull request by peckto**
Refactor code and add new functions `init_tensorflow()`, `calc_mask()` and data structure `calcinfo_t`.
In preparation for #38

### CODE REVIEW

File: deepseg.cc [hunk 1]
```diff
@@ -82,6 +82,9 @@ cv::Mat getTensorMat(int tnum, int debug) {
 
 // deeplabv3 classes
 std::vector<std::string> labels = { "background", "aeroplane", "bicycle", "bird", "boat", "bottle", "bus", "car", "cat", "chair", "cow", "dining table", "dog", "horse", "motorbike", "person", "potted plant", "sheep", "sofa", "train", "tv" };
+// label number of "person" for DeepLab v3+ model
+const int cnum = labels.size();
+const int pers = std::find(labels.begin(),labels.end(),"person") - labels.begin();
```
Comment (BenBE) [inline deepseg.cc:87]: Proper C++ type usually gives `size_t` for containers. Also error checking for `std::find` returning "no match found" is missing. Last Prefer `std::distance` for retrieving the index of the `pers` label. (Internally also maps to basically this, but allos the STL to optimize.

File: deepseg.cc [hunk 2]
```diff
@@ -92,6 +95,24 @@ typedef struct {
 	pthread_mutex_t lock;
 } capinfo_t;
 
+typedef struct {
+	const char *modelname;
+	int threads;
+	int width;
+	int height;
```
Comment (BenBE) [inline deepseg.cc:102]: Candidates for use of `size_t` …

File: deepseg.cc [hunk 3]
```diff
@@ -110,6 +131,111 @@ void *grab_thread(void *arg) {
...
+		out[n] = (val & 0xE0) | (out[n] >> 3);
+	}
+
+	// threshold probability
+	if (strstr(info->modelname,"body-pix"))
+	for (unsigned int n = 0; n < info->output.total(); n++) {
+		// FIXME: hardcoded threshold
+		uint8_t val = (tmp[n] > 0.65 ? 0 : 255);
+		out[n] = (val & 0xE0) | (out[n] >> 3);
+	}
```
Comment (BenBE) [inline deepseg.cc:211]: Misleading indentation. Please use braces and indent the for loop.

File: deepseg.cc [hunk 4]
```diff
@@ -110,6 +131,111 @@ void *grab_thread(void *arg) {
...
+		 * softmax across both channels to yield foreground
+		 * probability in [0.0, 1.0]. */
+	for (unsigned int n = 0; n < info->output.total(); n++) {
+		float exp0 = expf(tmp[2*n  ]);
+		float exp1 = expf(tmp[2*n+1]);
+		float p0 = exp0 / (exp0+exp1);
+		float p1 = exp1 / (exp0+exp1);
+		uint8_t val = (p0 < p1 ? 0 : 255);
+		out[n] = (val & 0xE0) | (out[n] >> 3);
+	}
```
Comment (BenBE) [inline deepseg.cc:228]: dito.

**Review by BenBE — CHANGES_REQUESTED** [review]

File: deepseg.cc (hunk 1 above)
Comment (peckto) [reply deepseg.cc]: I fixed the `size_t` and `std::distance`.
As the `labels` list is predefined, I would not expect `find` to fail.

File: deepseg.cc (hunk 2 above)
Comment (peckto) [reply deepseg.cc]: fixed

File: deepseg.cc (hunk 3 above)
Comment (peckto) [reply deepseg.cc]: fixed

File: deepseg.cc (hunk 4 above)
Comment (peckto) [reply deepseg.cc]: fixed

File: deepseg.cc [hunk 5]
```diff
@@ -110,16 +131,124 @@ void *grab_thread(void *arg) {
 	return NULL;
 }
 
+void init_tensorflow(calcinfo_t *info) {
```
Comment (BenBE) [inline deepseg.cc:134]: Probably should be reference, not pointer.

File: deepseg.cc [hunk 6]
```diff
@@ -110,16 +131,124 @@ void *grab_thread(void *arg) {
...
+	info->mroi = info->mask(info->roidim);
+
+	// erosion/dilation element
+	info->element = cv::getStructuringElement( cv::MORPH_RECT, cv::Size(5,5) );
+
+	// create Mat for small mask
+	info->ofinal = cv::Mat(info->output.rows,info->output.cols,CV_8UC1);
+}
+
+void calc_mask(calcinfo_t *info) {
```
Comment (BenBE) [inline deepseg.cc:171]: dito.

File: deepseg.cc [hunk 7]
```diff
@@ -164,23 +293,23 @@ int main(int argc, char* argv[]) {
 				showUsage = true;
 			}
 		} else if (strncmp(argv[arg], "-w", 2)==0) {
-			if (hasArgument && sscanf(argv[++arg], "%d", &width)) {
+			if (hasArgument && sscanf(argv[++arg], "%ld", &width)) {
```
Comment (BenBE) [inline deepseg.cc:296]: With `width`, `height`, and `threads` being `size_t` it probably should be `%zu` for the format code.
Check the `-Wformat=2` compiler option from `gcc` and `clang`

File: deepseg.cc [hunk 8]
```diff
@@ -164,23 +293,23 @@ int main(int argc, char* argv[]) {
...
 				if (!width) {
 					showUsage = true;
 				}
 			} else {
 				showUsage = true;
 			}
 		} else if (strncmp(argv[arg], "-h", 2)==0) {
-			if (hasArgument && sscanf(argv[++arg], "%d", &height)) {
+			if (hasArgument && sscanf(argv[++arg], "%ld", &height)) {
```
Comment (BenBE) [inline deepseg.cc:304]: dito.

File: deepseg.cc [hunk 9]
```diff
@@ -164,23 +293,23 @@ int main(int argc, char* argv[]) {
...
 				if (!height) {
 					showUsage = true;
 				}
 			} else {
 				showUsage = true;
 			}
 		} else if (strncmp(argv[arg], "-t", 2)==0) {
-			if (hasArgument && sscanf(argv[++arg], "%d", &threads)) {
+			if (hasArgument && sscanf(argv[++arg], "%ld", &threads)) {
```
Comment (BenBE) [inline deepseg.cc:312]: dito.

File: deepseg.cc [hunk 10]
```diff
@@ -213,11 +342,11 @@ int main(int argc, char* argv[]) {
...
 	printf("ccam:   %s\n", ccam);
 	printf("vcam:   %s\n", vcam);
-	printf("width:  %d\n", width);
-	printf("height: %d\n", height);
+	printf("width:  %ld\n", width);
+	printf("height: %ld\n", height);
 	printf("flip_h: %s\n", flipHorizontal ? "yes" : "no");
 	printf("flip_v: %s\n", flipVertical ? "yes" : "no");
-	printf("threads:%d\n", threads);
+	printf("threads:%ld\n", threads);
```
Comment (BenBE) [inline deepseg.cc:349]: dito.

File: deepseg.cc [hunk 11]
```diff
@@ -310,93 +402,30 @@ int main(int argc, char* argv[]) {
...
-			float p1 = exp1 / (exp0+exp1);
-			uint8_t val = (p0 < p1 ? 0 : 255);
-			out[n] = (val & 0xE0) | (out[n] >> 3);
-		}
-
-		// denoise
-		cv::Mat tmpbuf;
-		cv::dilate(ofinal,tmpbuf,element);
-		cv::erode(tmpbuf,ofinal,element);
+		calcinfo.raw = (*capinfo.raw);
```
Comment (BenBE) [inline deepseg.cc:405]: Unnecessary parens

**Review by BenBE — CHANGES_REQUESTED** [review]
Please rebase to current `experimental` or `main` branch to resolve merge conflicts.
Additionally please fixup the two minor bugfix commits into the main commit.
Lastly, please separate the code moving from any other code changes as this simplifies reviewing.

File: deepseg.cc (hunk 5 above)
Comment (peckto) [reply deepseg.cc]: done

File: deepseg.cc (hunk 6 above)
Comment (peckto) [reply deepseg.cc]: done

File: deepseg.cc (hunk 7 above)
Comment (peckto) [reply deepseg.cc]: done

File: deepseg.cc (hunk 8 above)
Comment (peckto) [reply deepseg.cc]: done

**Comment by peckto** [comment]
Thanks for your feedback. I solved your comments.
Does the new commit structure help you to review the changes?

File: deepseg.cc [hunk 12]
```diff
@@ -109,6 +109,9 @@ cv::Mat getTensorMat(int tnum, int debug) {
 
 // deeplabv3 classes
 std::vector<std::string> labels = { "background", "aeroplane", "bicycle", "bird", "boat", "bottle", "bus", "car", "cat", "chair", "cow", "dining table", "dog", "horse", "motorbike", "person", "potted plant", "sheep", "sofa", "train", "tv" };
```
Comment (BenBE) [inline deepseg.cc:111]: Probably make this `const`.

File: deepseg.cc [hunk 13]
```diff
@@ -377,6 +398,8 @@ int main(int argc, char* argv[]) {
 	printf("Startup: %ldns\n", diffnanosecs(ti.lastns,ti.bootns));
 
 	bool filterActive = true;
+	calcinfo_t calcinfo = { modelname, threads, width, height, debug };
+	init_tensorflow(calcinfo);
```
Comment (BenBE) [inline deepseg.cc:481]: These two lines don't compile in the context of this commit.

**Review by BenBE — CHANGES_REQUESTED** [review]
Thanks for the splitting of the commits. Makes things much easier to review.

Two things still left: First one is the `init_tensorflow` call doesn't compile as part of the first commit and one minor `const`-nes issue.

File: deepseg.cc (hunk 12 above)
Comment (peckto) [reply deepseg.cc]: fixed

File: deepseg.cc (hunk 13 above)
Comment (peckto) [reply deepseg.cc]: fixed

**Review by BenBE — APPROVED** [review]
Thx, LGTM.

**Comment by phlash** [comment]
Nicely done, thanks. LGTM too, @BenBE please merge!

**Pull request merged**